storage/src/desktop/storage_reference_desktop.cc

```diff
@@ -368,22 +469,34 @@ Future<Metadata> StorageReferenceInternal::PutFileInternal(
   auto* future_api = future();
   auto handle = future_api->SafeAlloc<Metadata>(kStorageReferenceFnPutFile);
 
-  // Open the file, calculate the length.
-  storage::internal::RequestFile* request(
-      new storage::internal::RequestFile(StripProtocol(path).c_str(), 0));
-  if (!request->IsFileOpen()) {
-    delete request;
-    future_api->Complete(handle, kErrorUnknown, "Could not read file.");
-  } else {
-    // Everything is good.  Fire off the request.
-    ReturnedMetadataResponse* response =
-        new ReturnedMetadataResponse(handle, future_api, AsStorageReference());
-
-    PrepareRequest(request, storageUri_.AsHttpUrl().c_str(), rest::util::kPost);
-    RestCall(request, request->notifier(), response, handle.get(), listener,
-             controller_out);
-  }
-
+  std::string final_path = StripProtocol(path);
+  auto send_request_funct{
+      [&, final_path, listener, controller_out]() -> BlockingResponse* {
+        auto* future_api = future();
+        auto handle =
+            future_api->SafeAlloc<Metadata>(kStorageReferenceFnPutFileInternal);
+
+        // Open the file, calculate the length.
+        storage::internal::RequestFile* request(
+            new storage::internal::RequestFile(final_path.c_str(), 0));
+        if (!request->IsFileOpen()) {
+          delete request;
+          future_api->Complete(handle, kErrorUnknown, "Could not read file.");
```

Comment:
PutFile has the unique failure case of being unable to read the file. Since this completes with an error, it should not have a retryable http status, and so it should not retry the request.

Reply:
Makes sense.